sway-core/src/lib.rs

```diff
@@ -319,22 +319,22 @@ pub fn compile_to_asm(
     build_config: BuildConfig,
 ) -> CompileResult<CompiledAsm> {
     let ast_res = compile_to_ast(type_engine, input, initial_namespace, Some(&build_config));
-    ast_to_asm(type_engine, ast_res, &build_config)
+    ast_to_asm(type_engine, &ast_res, &build_config)
 }
 
 /// Given an AST compilation result,
 /// try compiling to a `CompiledAsm`,
 /// containing the asm in opcode form (not raw bytes/bytecode).
 pub fn ast_to_asm(
     type_engine: &TypeEngine,
-    ast_res: CompileResult<ty::TyProgram>,
+    ast_res: &CompileResult<ty::TyProgram>,
```

Comment:
This is a bit weird, even prior to your change.  I wonder if we need to take a look at the flow through `sway-core/src/lib.rs` and reassess the API.  I think it's grown fairly organically over the months and needs some chopping.

Reply:
Yeah, I think a *slightly* more idiomatic way of writing this would be:
```suggestion
    ast_res: CompileResult<&ty::TyProgram>,
```
We might want to add an `as_ref` method to `CompileResult` if we don't already have one (similar to [`std::result::Result::as_ref`](https://doc.rust-lang.org/std/result/enum.Result.html#method.as_ref)).

All that said, this function probably shouldn't take a `CompileResult` as an argument anyway, and would likely be better off taking the `&ty::TyProgram` directly. But we can address that in the future :)